### PR TITLE
Feature/ah shell

### LINKED ISSUE #1435: Build app shell - Create and set background colour based on location
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### LINKED ISSUE #1436: Header - Integrate header component
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### LINKED ISSUE #1437: Add search bar (no functionality)
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### LINKED ISSUE #1438: add drawer and action buttons
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### LINKED ISSUE #1439: Add avatar - Header
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### PR DESCRIPTION
**Pull request by stevenbister**
### 🧹 Please check off the following before submitting:

- [x] Close issues
- [x] Request reviewers
- [x] Assign yourself
- [ ] Add labels (e.g. `build:osc-ecommerce`)
- [x] Add project
- [x] Set project status
- [x] Add milestone

---

## ☑ List all github issues that this PR will close:

- Closes #1436 
- Closes #1438 
- Closes #1439 
- Closes #1435
- Closes #1437

## 📝 Description

Adds a `_protected` layout route and an `AppShell` component to wrap around the protected pages in the academic hub. 
This includes the `Header` and the `Drawer` buttons for the notifications and the calendar. I've added a couple of placeholders: 
- the sidebar which can be added once #1470 is merged -- I have opened a ticket to handle this separately #1599
- ~~the search bar is the wrong variant can be updated by #1519~~
- ~~the background colours can't be associated to a user type yet and can be updated when #1398 goes in~~

I've also added a `Title` component to make headings consistent and less hassle to add.

## ⛳️ Current behaviour (updates)

N/a

## 🚀 New behaviour

- Adds protected layout route
- Adds app shell component
- Adds site header
- Adds title component
- Adds notification and calendar drawers

## 💣 Is this a breaking change (Yes/No): No

> Please describe any breaking changes.

## 🧾 Final Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible cross-device impacts (iOS, Android)
- [x] I've considered possible cross-browser impacts (Chrome, Safari, Edge, Firefox)
- [ ] I've added tests to cover my changes (Unit, E2E, Integration, Contract, Functional, Acceptance etc)
- [ ] I've added or updated any documentation.

## 📝 Additional Information

There's no avatar image at the moment, we need to decide where that's going to sit (cloudinary/AWS) so gonna hold off adding that until I know what the output will be.

I've changed from the design a bit to simplify how the background colours will work. In the design they look set to each user role, however users can have multiple roles (admin and tutor for example) so handling how to set the colour was getting tricky -- instead I've just done it based on the user location.

You will also need some new env vars and probably a user to test this now. Let me know when you look at it and I'll send you over the latest bits and pieces

### Summary by CodeRabbit

- New Feature: Introduced new drawer components `CalendarDrawer`, `NotificationDrawer`, and `ProfileDrawer` for improved navigation.
- New Feature: Added `SiteHeader`, `AppShell`, and `AppShellMain` components to enhance the layout of the application.
- Refactor: Replaced existing `<h1>` elements with a new `Title` component across various routes for consistent styling.
- Refactor: Updated the `LogoutForm` component to accept `LogoutFormProps` as props and spread the remaining props.
- New Feature: Added a `Protected` component for rendering protected pages with a site header, sidebar, and main content area.
- Chore: Performed code cleanup in the root component by removing unused imports and variables, updating import statements for CSS stylesheets, and reorganizing the code structure.
- Style: Updated various styling rules and added new import statements for styles. Also updated the `HeaderInnerProps` interface in the `Header` component.
- Style: Introduced new gradient color definitions and background color values for enhanced UI design.
- Refactor: Updated the `TextInput` component to accept `containerClassName` and `outerContainerClassName` properties for more flexible styling.

### CODE REVIEW

File: packages/osc-academic-hub/app/components/Header/Header.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        <Header variant="secondary" {...rest}>
+            <HeaderInner className="c-header__inner--single">
+                <Logo />
+
+                <HeaderActionBar>
+                    <TextInput
+                        label="Search"
+                        id="search"
+                        name="search"
+                        // TODO: Update to quinary variant once #1519 is merged
```
[the remark this answers is not in the document]
Comment (hi-im-si) [reply packages/osc-academic-hub/app/components/Header/Header.tsx:27]: @stevenbister #1519 is now closed so worth updating

File: packages/osc-academic-hub/app/routes/_protected.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    ) {
+        bgColour = BACKGROUND.secondary;
+    } else if (
+        location.pathname.includes(PATHS.STUDENT_LIST) ||
+        location.pathname.includes(PATHS.STUDENT_RECORD)
+    ) {
+        bgColour = BACKGROUND.tertiary;
+    } else {
+        bgColour = BACKGROUND.primary;
+    }
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/routes/_protected.tsx]: I like that actually, I've had to tweak it slightly so I can get the values of the `pathColorMap` out, your current example only return the properties

File: packages/osc-academic-hub/app/routes/_protected.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        [PATHS.STUDENT_RECORD]: BACKGROUND.tertiary,
+    };
+
+    // Set the background colour based on the route
+    const findBackgroundColor = Object.entries(pathColorMap).find(([key, _value]) =>
+        location.pathname.includes(key)
+    ) || ['other', BACKGROUND.primary];
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars -- we don't need the key
+    const [_, bgColour] = findBackgroundColor;
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/routes/_protected.tsx]: Ah yes you are right, that is much simpler!

File: packages/osc-ui/src/components/Header/header.scss
Comment (hi-im-si) [inline packages/osc-ui/src/components/Header/header.scss:1]: adding a `display: flex;` provides better alignment to the icons in the header. 
See images:

BEFORE:
image

AFTER (applying flex):
image

Was there a specific reason for it being block displayed?

**Review by hi-im-si — CHANGES_REQUESTED** [review]
literally the most minor of comments but its pretty much gtg!

File: packages/osc-ui/src/components/Header/header.scss
Comment (stevenbister) [reply packages/osc-ui/src/components/Header/header.scss:1]: I honestly can't remember lol, but yeah flex does look better. I'll check it doesn't mess up the ecommerce header and get that updated

**Review by hi-im-si — APPROVED** [review]

**Review by hi-im-si — APPROVED** [review]

**Pull request merged**